src/script/sign.h

```diff
@@ -22,13 +22,32 @@ struct CMutableTransaction;
 
 struct KeyOriginInfo
 {
-    unsigned char fingerprint[4];
+    unsigned char fingerprint[4]; // First 32 bits of the Hash160 of the public key at the root of the path
     std::vector<uint32_t> path;
 
     friend bool operator==(const KeyOriginInfo& a, const KeyOriginInfo& b)
     {
         return std::equal(std::begin(a.fingerprint), std::end(a.fingerprint), std::begin(b.fingerprint)) && a.path == b.path;
     }
+
+    ADD_SERIALIZE_METHODS;
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action)
+    {
+        READWRITE(fingerprint);
+        READWRITE(path);
+    }
+
+    void SetNull()
+    {
+        memset(fingerprint, 0, 4);
+        path.clear();
+    }
+
+    bool IsNull() const
+    {
+        return path.empty() && fingerprint[0] == 0 && fingerprint[1] == 0 && fingerprint[2] == 0 && fingerprint[3] == 0;
```

Comment:
:+1: Don't special-case values that can actually occur

Reply:
Can we mandate that `KeyOriginInfo` must always contain a `fingerprint`? So in that case justing checking `path.empty()` is enough.

Reply:
How else then would we determine that a KeyOriginInfo is null?

Reply:
So then it seems we need an explicit "null" for the fingerprint to indicate it's not set.

Reply:
Adding a `null` field would cause a lot of code churn. Instead I've changed `CKeyMetadata` to have a boolean that indicates whether the `key_origin` is valid. Let me know what you think of that; if it's good, I'll squash the commit down.

Reply:
Looks better to me